rs/consensus/vetkd/src/lib.rs

```diff
@@ -0,0 +1,397 @@
+            let size = NumBytes::new((accumulated_size + candidate_size) as u64);
+            if size >= max_payload_size {
+                break;
+            } else {
+                accumulated_size += candidate_size;
+                candidates.insert(*callback_id, candidate);
+            }
+        }
+
+        VetKdPayload {
+            vetkd_agreements: candidates,
+        }
+    }
+
+    fn validate_vetkd_payload_impl(
+        &self,
+        payload: VetKdPayload,
+        context_time: Time,
+        config: ChainKeyConfig,
+        state: &ReplicatedState,
+        delivered_ids: HashSet<CallbackId>,
+    ) -> Result<(), PayloadValidationError> {
+        let (valid_keys, request_expiry_time) = get_valid_keys_and_expiry(config, context_time);
+
+        let contexts = state.signature_request_contexts();
+
+        for (id, agreement) in payload.vetkd_agreements {
+            if delivered_ids.contains(&id) {
+                return invalid_artifact_err(InvalidVetKdPayloadReason::DuplicateResponse(id));
+            }
+
+            let Some(context) = contexts.get(&id) else {
+                return invalid_artifact_err(InvalidVetKdPayloadReason::MissingContext(id));
+            };
+
+            if !context.is_vetkd() {
+                return invalid_artifact_err(InvalidVetKdPayloadReason::IDkgContext(id));
+            }
+
+            let expected_reject = reject_if_invalid(&valid_keys, context, request_expiry_time);
+
+            match agreement {
+                VetKdAgreement::Success(data) => {
+                    if expected_reject.is_some() {
+                        return invalid_artifact_err(
+                            InvalidVetKdPayloadReason::MismatchedAgreement {
+                                expected: expected_reject,
+                                received: None,
+                            },
+                        );
+                    } else {
+                        self.validate_vetkd_agreement(id, context, data)?
+                    }
+                }
+                reject => {
+                    if Some(&reject) != expected_reject.as_ref() {
+                        return invalid_artifact_err(
+                            InvalidVetKdPayloadReason::MismatchedAgreement {
+                                expected: expected_reject,
+                                received: Some(reject),
+                            },
+                        );
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Validate the VetKD key given in `data` according to the context of its request.
+    fn validate_vetkd_agreement(
+        &self,
+        id: CallbackId,
+        context: &SignWithThresholdContext,
+        data: Vec<u8>,
+    ) -> Result<(), PayloadValidationError> {
+        let ThresholdArguments::VetKd(ctxt_args) = &context.args else {
+            return invalid_artifact_err(InvalidVetKdPayloadReason::IDkgContext(id));
```

Comment:
NIT: Would name the error `NotNiDkgContext`, since it is not clear from the error name, whats wrong with it.